File: code/go/detect.go
package main

import (
	"fmt"
	"os"
	"strings"
)

// readFirstLine reads the file at path via fs and returns the first non-empty
// line, trimmed of leading and trailing whitespace.  An empty string is
// returned when the file cannot be read or contains no non-empty lines.
func readFirstLine(fs Filesystem, path string) string {
	content, err := fs.ReadFile(path)
	if err != nil {
		return ""
	}
	for _, line := range strings.Split(content, "\n") {
		trimmed := strings.TrimSpace(line)
		if trimmed != "" {
			return trimmed
		}
	}
	return ""
}

// parseOsRelease parses a KEY=VALUE formatted file (as used by /etc/os-release)
// and returns a map of keys to values.  Surrounding double- or single-quote
// characters are stripped from each value.
func parseOsRelease(content string) map[string]string {
	result := make(map[string]string)
	for _, line := range strings.Split(content, "\n") {
		line = strings.TrimSpace(line)
		// Skip blank lines and comments.
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		idx := strings.IndexByte(line, '=')
		if idx < 0 {
			continue
		}
		key := strings.TrimSpace(line[:idx])
		val := strings.TrimSpace(line[idx+1:])
		// Strip a single layer of matching outer quotes (double or single).
		if len(val) >= 2 {
			if (val[0] == '"' && val[len(val)-1] == '"') ||
				(val[0] == '\'' && val[len(val)-1] == '\'') {
				val = val[1 : len(val)-1]
			}
		}
		result[key] = val
	}
	return result
}

// detectDistribution probes well-known release files via fs and returns a
// DistroInfo describing the detected Linux distribution family, release
// string, and package-management backend.
//
// Detection order (first match wins):
//  1. /etc/debian_version      → deb / dpkg
//  2. /etc/redhat-release      → rpm / /bin/rpm
//  3. /etc/UnitedLinux-release AND /etc/SuSE-release → rpm / /bin/rpm (combined release)
//  4. /etc/UnitedLinux-release → rpm / /bin/rpm
//  5. /etc/SLOX-release        → rpm / /bin/rpm
//  6. /etc/SuSE-release        → rpm / /bin/rpm
//  7. /etc/os-release          → rpm / /usr/bin/rpm  (PRETTY_NAME value)
//  8. unknown / none           → warning emitted to stderr
func detectDistribution(fs Filesystem) DistroInfo {
	// ── Step 1: Debian / Ubuntu ───────────────────────────────────────────────
	if fs.Exists("/etc/debian_version") {
		return DistroInfo{
			Family:     FamilyDeb,
			Release:    readFirstLine(fs, "/etc/debian_version"),
			Backend:    BackendDpkg,
			DpkgStatus: "/var/lib/dpkg/status",
		}
	}

	// ── Step 2: Red Hat / CentOS / Fedora / RHEL ─────────────────────────────
	if fs.Exists("/etc/redhat-release") {
		return DistroInfo{
			Family:  FamilyRPM,
			Release: readFirstLine(fs, "/etc/redhat-release"),
			Backend: BackendRPM,
			RpmCmd:  "/bin/rpm",
		}
	}

	// ── Step 3: UnitedLinux + SuSE (both present) ────────────────────────────
	if fs.Exists("/etc/UnitedLinux-release") && fs.Exists("/etc/SuSE-release") {
		ulLine := readFirstLine(fs, "/etc/UnitedLinux-release")
		suseLine := readFirstLine(fs, "/etc/SuSE-release")
		return DistroInfo{
			Family:  FamilyRPM,
			Release: ulLine + ", " + suseLine,
			Backend: BackendRPM,
			RpmCmd:  "/bin/rpm",
		}
	}

	// ── Step 4: UnitedLinux only ──────────────────────────────────────────────
	if fs.Exists("/etc/UnitedLinux-release") {
		return DistroInfo{
			Family:  FamilyRPM,
			Release: readFirstLine(fs, "/etc/UnitedLinux-release"),
			Backend: BackendRPM,
			RpmCmd:  "/bin/rpm",
		}
	}

	// ── Step 5: SLOX ──────────────────────────────────────────────────────────
	if fs.Exists("/etc/SLOX-release") {
		return DistroInfo{
			Family:  FamilyRPM,
			Release: readFirstLine(fs, "/etc/SLOX-release"),
			Backend: BackendRPM,
			RpmCmd:  "/bin/rpm",
		}
	}

	// ── Step 6: SuSE only ────────────────────────────────────────────────────
	if fs.Exists("/etc/SuSE-release") {
		return DistroInfo{
			Family:  FamilyRPM,
			Release: readFirstLine(fs, "/etc/SuSE-release"),
			Backend: BackendRPM,
			RpmCmd:  "/bin/rpm",
		}
	}

	// ── Step 7: systemd /etc/os-release (modern fallback) ────────────────────
	if fs.Exists("/etc/os-release") {
		content, err := fs.ReadFile("/etc/os-release")
		release := ""
		if err == nil {
			fields := parseOsRelease(content)
			release = fields["PRETTY_NAME"]
		}
		return DistroInfo{
			Family:  FamilyRPM,
			Release: release,
			Backend: BackendRPM,
			RpmCmd:  "/usr/bin/rpm",
		}
	}

	// ── Step 8: Unknown ───────────────────────────────────────────────────────
	fmt.Fprintln(os.Stderr, "distribution not supported")
	return DistroInfo{
		Family:  FamilyUnknown,
		Backend: BackendNone,
	}
}

File: code/rs/pcd-sonnet-output/src/detect.rs
// detect.rs — detect-distribution BEHAVIOR
#![allow(dead_code)]

use crate::interfaces::Filesystem;
use crate::types::{DistributionFamily, DistributionInfo, PackageVersioningBackend};

/// Determine the Linux distribution family and select the appropriate
/// package versioning backend. STEPS order is the precedence order.
pub fn detect_distribution(fs: &dyn Filesystem) -> DistributionInfo {
    // Step 1: Debian family
    if fs.exists("/etc/debian_version") {
        let release = fs.read_file("/etc/debian_version")
            .unwrap_or_default()
            .lines()
            .next()
            .unwrap_or("")
            .trim()
            .to_string();
        return DistributionInfo {
            family:      DistributionFamily::Deb,
            release,
            backend:     PackageVersioningBackend::Dpkg,
            rpm_cmd:     String::new(),
            dpkg_status: "/var/lib/dpkg/status".to_string(),
        };
    }

    // Step 2: Red Hat family
    if fs.exists("/etc/redhat-release") {
        let release = fs.read_file("/etc/redhat-release")
            .unwrap_or_default()
            .lines()
            .next()
            .unwrap_or("")
            .trim()
            .to_string();
        return DistributionInfo {
            family:      DistributionFamily::Rpm,
            release,
            backend:     PackageVersioningBackend::Rpm,
            rpm_cmd:     "/bin/rpm".to_string(),
            dpkg_status: String::new(),
        };
    }

    // Step 3: UnitedLinux AND SuSE-release
    if fs.exists("/etc/UnitedLinux-release") && fs.exists("/etc/SuSE-release") {
        let ul = fs.read_file("/etc/UnitedLinux-release")
            .unwrap_or_default()
            .lines()
            .next()
            .unwrap_or("")
            .trim()
            .to_string();
        let suse = fs.read_file("/etc/SuSE-release")
            .unwrap_or_default()
            .lines()
            .next()
            .unwrap_or("")
            .trim()
            .to_string();
        let release = format!("{}, {}", ul, suse);
        return DistributionInfo {
            family:      DistributionFamily::Rpm,
            release,
            backend:     PackageVersioningBackend::Rpm,
            rpm_cmd:     "/bin/rpm".to_string(),
            dpkg_status: String::new(),
        };
    }

    // Step 4: UnitedLinux only
    if fs.exists("/etc/UnitedLinux-release") {
        let release = fs.read_file("/etc/UnitedLinux-release")
            .unwrap_or_default()
            .lines()
            .next()
            .unwrap_or("")
            .trim()
            .to_string();
        return DistributionInfo {
            family:      DistributionFamily::Rpm,
            release,
            backend:     PackageVersioningBackend::Rpm,
            rpm_cmd:     "/bin/rpm".to_string(),
            dpkg_status: String::new(),
        };
    }

    // Step 5: SLOX
    if fs.exists("/etc/SLOX-release") {
        let release = fs.read_file("/etc/SLOX-release")
            .unwrap_or_default()
            .lines()
            .next()
            .unwrap_or("")
            .trim()
            .to_string();
        return DistributionInfo {
            family:      DistributionFamily::Rpm,
            release,
            backend:     PackageVersioningBackend::Rpm,
            rpm_cmd:     "/bin/rpm".to_string(),
            dpkg_status: String::new(),
        };
    }

    // Step 6: SuSE-release
    if fs.exists("/etc/SuSE-release") {
        let release = fs.read_file("/etc/SuSE-release")
            .unwrap_or_default()
            .lines()
            .next()
            .unwrap_or("")
            .trim()
            .to_string();
        return DistributionInfo {
            family:      DistributionFamily::Rpm,
            release,
            backend:     PackageVersioningBackend::Rpm,
            rpm_cmd:     "/bin/rpm".to_string(),
            dpkg_status: String::new(),
        };
    }

    // Step 7: /etc/os-release
    if fs.exists("/etc/os-release") {
        let content = fs.read_file("/etc/os-release").unwrap_or_default();
        let release = parse_os_release_field(&content, "PRETTY_NAME")
            .unwrap_or_else(|| "Linux".to_string());
        return DistributionInfo {
            family:      DistributionFamily::Rpm,
            release,
            backend:     PackageVersioningBackend::Rpm,
            rpm_cmd:     "/usr/bin/rpm".to_string(),
            dpkg_status: String::new(),
        };
    }

    // Step 8: unknown
    eprintln!("sitar: distribution not supported");
    DistributionInfo {
        family:      DistributionFamily::Unknown,
        release:     String::new(),
        backend:     PackageVersioningBackend::None,
        rpm_cmd:     String::new(),
        dpkg_status: String::new(),
    }
}

/// Parse a KEY=VALUE or KEY="VALUE" line from /etc/os-release content.
pub fn parse_os_release_field(content: &str, key: &str) -> Option<String> {
    for line in content.lines() {
        if let Some(rest) = line.strip_prefix(key) {
            if let Some(val) = rest.strip_prefix('=') {
                let v = val.trim().trim_matches('"').to_string();
                if !v.is_empty() {
                    return Some(v);
                }
            }
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::interfaces::FakeFilesystem;

    #[test]
    fn test_detect_debian() {
        let mut fs = FakeFilesystem::new();
        fs.files.insert("/etc/debian_version".to_string(), "12.0\n".to_string());
        let info = detect_distribution(&fs);
        assert_eq!(info.family, DistributionFamily::Deb);
        assert_eq!(info.backend, PackageVersioningBackend::Dpkg);
        assert_eq!(info.release, "12.0");
    }

    #[test]
    fn test_detect_redhat() {
        let mut fs = FakeFilesystem::new();
        fs.files.insert("/etc/redhat-release".to_string(), "Red Hat Enterprise Linux 9\n".to_string());
        let info = detect_distribution(&fs);
        assert_eq!(info.family, DistributionFamily::Rpm);
        assert_eq!(info.backend, PackageVersioningBackend::Rpm);
    }

    #[test]
    fn test_detect_os_release() {
        let mut fs = FakeFilesystem::new();
        fs.files.insert("/etc/os-release".to_string(),
            "NAME=\"openSUSE Leap\"\nPRETTY_NAME=\"openSUSE Leap 15.5\"\n".to_string());
        let info = detect_distribution(&fs);
        assert_eq!(info.family, DistributionFamily::Rpm);
        assert_eq!(info.release, "openSUSE Leap 15.5");
    }

    #[test]
    fn test_detect_unknown() {
        let fs = FakeFilesystem::new();
        let info = detect_distribution(&fs);
        assert_eq!(info.family, DistributionFamily::Unknown);
    }
}
